File: dotnet/src/Microsoft.Agents.Protocol.Runtime/Tools/ToolSchemaGenerator.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Reflection;

namespace Microsoft.Agents.Protocol.Runtime.Tools;

/// <summary>
/// Generates JSON schemas from .NET methods using reflection.
/// Shared utility across Client and Hosting SDKs for consistent schema generation.
/// </summary>
public static class ToolSchemaGenerator
{
    /// <summary>
    /// Generates a JSON schema from a delegate's parameters.
    /// </summary>
    /// <param name="handler">The delegate to analyze</param>
    /// <returns>JSON schema object</returns>
    public static object GenerateSchema(Delegate handler)
    {
        if (handler == null)
            throw new ArgumentNullException(nameof(handler));

        var method = handler.Method;
        var parameters = method.GetParameters();

        var properties = new Dictionary<string, object>();
        var required = new List<string>();

        foreach (var param in parameters)
        {
            if (param.Name == null) continue;

            properties[param.Name] = new
            {
                type = GetJsonType(param.ParameterType),
                description = $"Parameter {param.Name}"
            };

            if (!param.IsOptional && !param.HasDefaultValue)
            {
                required.Add(param.Name);
            }
        }

        return new
        {
            type = "object",
            properties,
            required
        };
    }

    /// <summary>
    /// Maps .NET types to JSON schema types.
    /// </summary>
    private static string GetJsonType(Type type)
    {
        // Handle nullable types
        var underlyingType = Nullable.GetUnderlyingType(type) ?? type;

        // Handle enums
        if (underlyingType.IsEnum)
            return "string";

        // Handle arrays and collections
        if (underlyingType.IsArray ||
            (underlyingType.IsGenericType && underlyingType.GetGenericTypeDefinition() == typeof(List<>)))
            return "array";

        // Map primitive types
        return Type.GetTypeCode(underlyingType) switch
        {
            TypeCode.String => "string",
            TypeCode.Int16 or TypeCode.Int32 or TypeCode.Int64 or
            TypeCode.UInt16 or TypeCode.UInt32 or TypeCode.UInt64 or
            TypeCode.Byte or TypeCode.SByte => "integer",
            TypeCode.Single or TypeCode.Double or TypeCode.Decimal => "number",
            TypeCode.Boolean => "boolean",
            TypeCode.DateTime => "string",
            _ => "object"
        };
    }
}

File: typescript/packages/agents-protocol-runtime/src/tools/ToolSchemaGenerator.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import type { JSONSchema } from '@microsoft/agents-protocol-abstractions';

/**
 * Generates JSON schemas from TypeScript functions.
 * Shared utility across Client and Hosting SDKs for consistent schema generation.
 *
 * NOTE: TypeScript has limited runtime reflection, so schemas must be provided
 * explicitly. This class validates and normalizes the provided schemas.
 *
 * @example
 * ```typescript
 * const schema = ToolSchemaGenerator.createObjectSchema({
 *   name: { type: 'string', description: 'The person name' },
 *   age: { type: 'number', description: 'The person age' }
 * }, ['name', 'age']);
 *
 * // schema.type === 'object'
 * // schema.required === ['name', 'age']
 * ```
 */
export class ToolSchemaGenerator {
  /**
   * Creates a simple object schema with typed properties.
   *
   * @param properties - Object properties with their schemas
   * @param required - Optional array of required property names
   * @returns Validated JSONSchema
   *
   * @example
   * ```typescript
   * const schema = ToolSchemaGenerator.createObjectSchema({
   *   query: { type: 'string', description: 'Search query' },
   *   limit: { type: 'number', description: 'Max results', default: 10 }
   * }, ['query']);
   * ```
   */
  static createObjectSchema(
    properties: Record<string, JSONSchema>,
    required?: string[]
  ): JSONSchema {
    return {
      type: 'object',
      properties,
      required: required || []
    };
  }

  /**
   * Creates a schema for a string parameter.
   *
   * @param description - Parameter description
   * @param options - Optional string constraints
   * @returns JSONSchema for string
   *
   * @example
   * ```typescript
   * const schema = ToolSchemaGenerator.createStringSchema(
   *   'User email address',
   *   { pattern: '^[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,}$' }
   * );
   * ```
   */
  static createStringSchema(
    description: string,
    options?: {
      minLength?: number;
      maxLength?: number;
      pattern?: string;
      enum?: string[];
    }
  ): JSONSchema {
    return {
      type: 'string',
      description,
      ...options
    };
  }

  /**
   * Creates a schema for a number parameter.
   *
   * @param description - Parameter description
   * @param options - Optional number constraints
   * @returns JSONSchema for number
   *
   * @example
   * ```typescript
   * const schema = ToolSchemaGenerator.createNumberSchema(
   *   'User age',
   *   { minimum: 0, maximum: 120 }
   * );
   * ```
   */
  static createNumberSchema(
    description: string,
    options?: {
      minimum?: number;
      maximum?: number;
      type?: 'integer' | 'number';
    }
  ): JSONSchema {
    return {
      type: options?.type || 'number',
      description,
      minimum: options?.minimum,
      maximum: options?.maximum
    };
  }

  /**
   * Creates a schema for a boolean parameter.
   *
   * @param description - Parameter description
   * @returns JSONSchema for boolean
   *
   * @example
   * ```typescript
   * const schema = ToolSchemaGenerator.createBooleanSchema('Enable notifications');
   * ```
   */
  static createBooleanSchema(description: string): JSONSchema {
    return {
      type: 'boolean',
      description
    };
  }

  /**
   * Creates a schema for an array parameter.
   *
   * @param description - Parameter description
   * @param items - Schema for array items
   * @param options - Optional array constraints
   * @returns JSONSchema for array
   *
   * @example
   * ```typescript
   * const schema = ToolSchemaGenerator.createArraySchema(
   *   'List of tags',
   *   { type: 'string' },
   *   { minItems: 1, maxItems: 10 }
   * );
   * ```
   */
  static createArraySchema(
    description: string,
    items: JSONSchema,
    options?: {
      minItems?: number;
      maxItems?: number;
    }
  ): JSONSchema {
    return {
      type: 'array',
      description,
      items,
      minItems: options?.minItems,
      maxItems: options?.maxItems
    };
  }

  /**
   * Validates a JSON schema.
   *
   * @param schema - The schema to validate
   * @throws Error if schema is invalid
   *
   * @example
   * ```typescript
   * const schema = { type: 'object', properties: {} };
   * ToolSchemaGenerator.validateSchema(schema); // OK
   *
   * const invalid = { properties: {} }; // Missing type
   * ToolSchemaGenerator.validateSchema(invalid); // Throws
   * ```
   */
  static validateSchema(schema: JSONSchema): void {
    if (!schema.type) {
      throw new Error('Schema must have a type field');
    }

    if (schema.type === 'object' && !schema.properties) {
      throw new Error('Object schema must have properties');
    }

    if (schema.type === 'array' && !schema.items) {
      throw new Error('Array schema must have items');
    }

    // Validate nested schemas
    if (schema.properties) {
      for (const [key, value] of Object.entries(schema.properties)) {
        try {
          ToolSchemaGenerator.validateSchema(value);
        } catch (error) {
          throw new Error(`Invalid schema for property '${key}': ${(error as Error).message}`);
        }
      }
    }

    if (schema.items && typeof schema.items === 'object' && !Array.isArray(schema.items)) {
      ToolSchemaGenerator.validateSchema(schema.items);
    }
  }

  /**
   * Merges two schemas (useful for extending base schemas).
   *
   * @param base - Base schema
   * @param extension - Schema to merge in
   * @returns Merged schema
   *
   * @example
   * ```typescript
   * const base = {
   *   type: 'object',
   *   properties: { name: { type: 'string' } },
   *   required: ['name']
   * };
   *
   * const extension = {
   *   properties: { age: { type: 'number' } },
   *   required: ['age']
   * };
   *
   * const merged = ToolSchemaGenerator.mergeSchemas(base, extension);
   * // merged.properties has both name and age
   * // merged.required includes both name and age
   * ```
   */
  static mergeSchemas(base: JSONSchema, extension: Partial<JSONSchema>): JSONSchema {
    return {
      ...base,
      ...extension,
      properties: {
        ...base.properties,
        ...extension.properties
      },
      required: [
        ...(base.required || []),
        ...(extension.required || [])
      ]
    };
  }
}
